pkg/controller/cert_control.go

```diff
@@ -0,0 +1,256 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package controller
+
+import (
+	"encoding/pem"
+	"fmt"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/pingcap/tidb-operator/pkg/label"
+	certutil "github.com/pingcap/tidb-operator/pkg/util/crypto"
+	capi "k8s.io/api/certificates/v1beta1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	types "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/kubernetes"
+	certlisters "k8s.io/client-go/listers/certificates/v1beta1"
+)
+
+// TiDBClusterCertOptions contains information needed to create new certificates
+type TiDBClusterCertOptions struct {
+	Namespace  string
+	Instance   string
+	CommonName string
+	HostList   []string
+	IPList     []string
+	Suffix     string
+	Component  string
+}
+
+// CertControlInterface manages certificates used by TiDB clusters
+type CertControlInterface interface {
+	Create(certOpts *TiDBClusterCertOptions) error
+	CheckSecret(ns string, secretName string) bool
+	//RevokeCert() error
+	//RenewCert() error
+}
+
+type realCertControl struct {
+	kubeCli    kubernetes.Interface
+	csrLister  certlisters.CertificateSigningRequestLister
+	secControl SecretControlInterface
+}
+
+// NewRealCertControl creates a new CertControlInterface
+func NewRealCertControl(
+	kubeCli kubernetes.Interface,
+	csrLister certlisters.CertificateSigningRequestLister,
+	secControl SecretControlInterface,
+) CertControlInterface {
+	return &realCertControl{
+		kubeCli:    kubeCli,
+		csrLister:  csrLister,
+		secControl: secControl,
+	}
+}
+
+func (rcc *realCertControl) Create(certOpts *TiDBClusterCertOptions) error {
+	var csrName string
+	if certOpts.Suffix == "" {
+		csrName = certOpts.Instance
+	} else {
+		csrName = fmt.Sprintf("%s-%s", certOpts.Instance, certOpts.Suffix)
+	}
+
+	// generate certificate if not exist
+	if rcc.secControl.Check(certOpts.Namespace, csrName) {
+		glog.Infof("Secret %s already exist, reusing the key pair. TidbCluster: %s/%s", csrName, certOpts.Namespace, csrName)
+		return nil
+	}
+
+	rawCSR, key, err := certutil.NewCSR(certOpts.CommonName, certOpts.HostList, certOpts.IPList)
+	if err != nil {
+		return fmt.Errorf("fail to generate new key and certificate for %s/%s, %v", certOpts.Namespace, csrName, err)
+	}
+
+	// sign certificate
+	csr, err := rcc.sendCSR(certOpts.Namespace, certOpts.Instance, rawCSR, csrName)
+	if err != nil {
+		return err
+	}
+	err = rcc.approveCSR(csr)
+	if err != nil {
+		return err
+	}
+
+	// wait at most 5min for the cert to be signed
+	timeout := int64(time.Minute.Seconds() * 5)
+	tick := time.After(time.Second * 10)
+	watchReq := types.ListOptions{
+		Watch:          true,
+		TimeoutSeconds: &timeout,
+		FieldSelector:  fields.OneTermEqualSelector("metadata.name", csrName).String(),
+	}
+
+	csrCh, err := rcc.kubeCli.Certificates().CertificateSigningRequests().Watch(watchReq)
```

Comment:
this will block the loop 5 minutes

Reply:
If everything work normal, the loop will exit at line 142 (usually within few seconds), the 5 min timeout is the worst case that CSR is not properly signed by k8s.